Torchrec enablement fixes - Itertools accumulate, hasattr w/o source

voznesenskym opened the pull request:
- support for itertools.accumulate
- https://github.com/pytorch/pytorch/issues/107690

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @gujinghui @PenghuiCheng @jianyuh @min-jean-cho @yanbing-j @Guobing-Chen @Xia-Weiwen @mcarilli @ptrblck @leslie-fang-intel @EikanWang @penguinwu @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @anijain2305

peterbell10 committed:
Remove some unnecessary <iostream> includes from headers (#106914)

In almost all cases this is only included for writing the output formatter, which
only uses `std::ostream` so including `<ostream>` is sufficient.

The istream header is ~1000 lines so the difference is non-trivial.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/106914
Approved by: https://github.com/lezcano

peterbell10 committed:
[ATen] Update pre-compiled header (#106915)

Pull Request resolved: https://github.com/pytorch/pytorch/pull/106915
Approved by: https://github.com/lezcano
ghstack dependencies: #106914

peterbell10 committed:
[inductor] WeakDep should not prevent dead node elimination (#107813)

A WeakDep is classed as a read dependency but the buffer is never actually read.
Instead it only effects schedule ordering. So for the purposes of dead node
elimination we should ignore WeakDeps.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/107813
Approved by: https://github.com/jansel, https://github.com/mlazos

XiaobingSuper committed:
inductor: support masked load for cpu path (#107670)

For max_pooling code:

```

#pragma GCC ivdep
                    for(long i2=static_cast<long>(0L); i2<static_cast<long>(56L); i2+=static_cast<long>(1L))
                    {
                        for(long i3=static_cast<long>(0L); i3<static_cast<long>(64L); i3+=static_cast<long>(16L))
                        {
                            auto tmp0 = at::vec::Vectorized<int>(static_cast<int>((-1L) + (2L*i1)));
                            auto tmp1 = at::vec::Vectorized<int>(static_cast<int>(0));
                            auto tmp2 = to_float_mask(tmp0 >= tmp1);
                            auto tmp3 = at::vec::Vectorized<int>(static_cast<int>(112));
                            auto tmp4 = to_float_mask(tmp0 < tmp3);
                            auto tmp5 = tmp2 & tmp4;
                            auto tmp6 = at::vec::Vectorized<int>(static_cast<int>((-1L) + (2L*i2)));
                            auto tmp7 = to_float_mask(tmp6 >= tmp1);
                            auto tmp8 = to_float_mask(tmp6 < tmp3);
                            auto tmp9 = tmp7 & tmp8;
                            auto tmp10 = tmp5 & tmp9;
                            auto tmp11 = [&]
                            {
                                auto tmp12 = at::vec::Vectorized<bfloat16>::loadu(in_ptr0 + static_cast<long>((-7232L) + i3 + (128L*i2) + (14336L*i1) + (802816L*i0)), 16);
                                                        load
                                auto tmp13 = cvt_lowp_fp_to_fp32<bfloat16>(tmp12);

                                return tmp13;
                            }
                            ;
                            auto tmp14 = decltype(tmp11())::blendv(at::vec::Vectorized<float>(-std::numeric_limits<float>::infinity()), tmp11(), to_float_mask(tmp10));
```

the index of ```tmp12 ``` may be a correct index, such as ```i1=0, i2=0, i3=0```, the index is ```-7232L```, it is not a valid index. We may meet segmentation fault error when we call ```tmp11()```, the original behavior is that only the ```tmp10```(index check variable) is true, we can safely get the value, this PR will support masked_load to fixing this issue.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/107670
Approved by: https://github.com/jgong5, https://github.com/jansel

xmfan committed:
Enhance fakepg: add fsdp+tp tests (#107626)

from working on a starter task with @wanchaol (T161350434):
Add the previously unsupported fsdp+tp example in FakePG, which required scatter and broadcast, as a unit test: https://github.com/pytorch/examples/blob/main/distributed/tensor_parallelism/two_d_parallel_example.py
Pull Request resolved: https://github.com/pytorch/pytorch/pull/107626
Approved by: https://github.com/wanchaol
ghstack dependencies: #107625

voznesenskym committed:
WIP

WIP

WIP

WIP

WIP

WIP

WIP

WIP

Fix

Update distributed.py

Fix

Commit:
Avoid decomposing `_unsafe_index` in Inductor (#107882)

`_unsafe_index` was previously added to the core ATen decomp table in https://github.com/pytorch/pytorch/pull/106814, but this has performance ramifications for Inductor. Therefore, this diff removes it from the decomposition table used by Inductor.

Differential Revision: [D48649210](https://our.internmc.facebook.com/intern/diff/D48649210/)

Pull Request resolved: https://github.com/pytorch/pytorch/pull/107882
Approved by: https://github.com/SherlockNoMad

wconstab committed:
Fix how DDPOptimizer clones dynamo callback (#107834)

Instead of hardcoding a new callback creation using 'convert_frame',
add an attribute to both callbacks that implement 'self cloning with new
backend', so DDPOptimizer can invoke this in a consistent way.

Fixes #107686

Pull Request resolved: https://github.com/pytorch/pytorch/pull/107834
Approved by: https://github.com/ezyang

anijain2305 committed:
[inductor][ac] preserve recompute tags through pattern matching (#107742)

Pull Request resolved: https://github.com/pytorch/pytorch/pull/107742
Approved by: https://github.com/eellison

BowenBao committed:
[ONNX] Enable 'ExportOutput.save' for models larger than 2GB (#107904)

Previously it fails during serialization, despite onnxscript graph_building managed to return ModelProto > 2GB.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/107904
Approved by: https://github.com/abock

jackiexu77 committed:
[inductor][fx passes] batch tanh in pre grad (#107881)

Summary:
Daohang report this pattern in f469463749
{F1074472207}
 {F1074473348}
Hence, we can fuse the tanh after same split.
Typically the pattern looks like split->getitem0,...n-> tanh(geitem 0,..., n). Hence, we search for parent node of tahn nodes and the node should be getitem(parent, index). If tanh is after same split node, parent nodes of getitem nodes should be same.

Test Plan:
```
[jackiexu0313@devgpu005.cln5 ~/fbsource/fbcode (c78736187)]$ buck test mode/dev-nosan //caffe2/test/inductor:group_batch_fusion
File changed: fbcode//caffe2/test/inductor/test_group_batch_fusion.py
Buck UI: https://www.internalfb.com/buck2/df87affc-d294-4663-a50d-ebb71b98070d
Test UI: https://www.internalfb.com/intern/testinfra/testrun/9570149208311124
Network: Up: 0B  Down: 0B
Jobs completed: 16. Time elapsed: 1:19.9s.
Tests finished: Pass 6. Fail 0. Fatal 0. Skip 0. Build failure 0
```

Differential Revision: D48581140

Pull Request resolved: https://github.com/pytorch/pytorch/pull/107881
Approved by: https://github.com/yanboliang

avikchaudhuri committed:
improved error message for IO mismatch (#107907)

Previously when we found some input or output mismatch between original args / traced result vs. graph-captured input / output, we would have a pretty sparse error message. (This might be partly due to the urge to reuse the same code for matching both inputs and outputs.)

With this PR we now point out which input or output is problematic, what its type is, and also present the expected types along with descriptions of what they mean. We don't suggest any fixes, but the idea is that it should be evident what went wrong looking at the error message.

Differential Revision: [D48668059](https://our.internmc.facebook.com/intern/diff/D48668059/)

Pull Request resolved: https://github.com/pytorch/pytorch/pull/107907
Approved by: https://github.com/gmagogsfm

malfet committed:
[MPS] Implement `polar` via metal shader (#107324)

Use `view_as_real` to cast complex into a pair of floats and then it becomes just another binary operator.

Enable `polar` and `view_as_complex` consistency tests, but skip `test_output_grad_match_polar_cpu` as `mul` operator is yet not supported

Remove redundant `#ifdef __OBJC__` and capture and re-throw exceptions captured during `createCacheBlock` block.
Fixes https://github.com/pytorch/pytorch/issues/78503

TODOs(in followup PRs):
  - Implement backwards (requires complex mul and sgn)
  - Measure the perf impact of computing the strides on the fly rather than ahead of time (unrelated to this PR)

Partially addresses https://github.com/pytorch/pytorch/issues/105665
Pull Request resolved: https://github.com/pytorch/pytorch/pull/107324
Approved by: https://github.com/albanD

pytorchupdatebot committed:
[vision hash update] update the pinned vision hash (#107911)

This PR is auto-generated nightly by [this action](https://github.com/pytorch/pytorch/blob/main/.github/workflows/_update-commit-hash.yml).
Update the pinned vision hash.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/107911
Approved by: https://github.com/pytorchbot

fffrog committed:
Fix the example of torch.slice_scatter (#107849)

Fixes #107681
fix the example of torch.slice_scatter
Pull Request resolved: https://github.com/pytorch/pytorch/pull/107849
Approved by: https://github.com/drisspg

jerryzh168 committed:
[quant][pt2e] Make sure XNNPACKQuantizer works with the pre_dispatch=True (#107872)

Summary: att

Test Plan:
```
buck test //executorch/backends/xnnpack/test:test_xnnpack_quantized_models -- test_resnet18

buck2 test 'fbcode//mode/opt' fbcode//caffe2/test:quantization_pt2e
```

Reviewed By: andrewor14, tugsbayasgalan

Differential Revision: D48415977

Pull Request resolved: https://github.com/pytorch/pytorch/pull/107872
Approved by: https://github.com/andrewor14

cccclai committed:
add get buffer from exported program (#107809)

Summary: We have the util function to get params, for parity we also need util function to get buffer`

Test Plan:
```
buck test //caffe2/test:test_export
```

Differential Revision: D48610877

Pull Request resolved: https://github.com/pytorch/pytorch/pull/107809
Approved by: https://github.com/JacobSzwejbka

gmagogsfm committed:
Expose torch.export.{save,load} APIs (#107888)

Pull Request resolved: https://github.com/pytorch/pytorch/pull/107888
Approved by: https://github.com/angelayi

anijain2305 committed:
[activation checkpointing] Add default autocast keys to functional rng wrappers (#107934)

Pull Request resolved: https://github.com/pytorch/pytorch/pull/107934
Approved by: https://github.com/xw285cornell

cpuhrsch committed:
[core][sparse][pruning] cuSPARSELt Kernels and ops. (#107398)

Summary:
This is a duplicate PR of 102133, which was reverted because it was
failing internal tests.

It seems like that internal builds did not like my guard to check if
cuSPARSELt was available or not.

Test Plan: python test/test_sparse_semi_structured.py

Differential Revision: D48440330

Pull Request resolved: https://github.com/pytorch/pytorch/pull/107398
Approved by: https://github.com/cpuhrsch

pearu committed:
Add optional is_coalesced argument to sparse coo tensor factory function. (#107638)

Resolves https://github.com/pytorch/pytorch/issues/107097

After this PR, instead of
```python
torch.sparse_coo_tensor(indices, values, size)._coalesced_(is_coalesced)
```
(that does not work in the autograd context, see #107097), use
```python
torch.sparse_coo_tensor(indices, values, size, is_coalesced=is_coalesced)
```

All sparse coo factory functions that take indices as input support the `is_coalesced` argument.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/107638
Approved by: https://github.com/cpuhrsch

voznesenskym commented:
> [<img alt="CLA Missing ID" width="328" height="28" src="https://camo.githubusercontent.com/10030abc9a0bd659c8175562a2f1a9f7786dcd921f5a5f454e6268ec7953fe17/68747470733a2f2f73332e616d617a6f6e6177732e636f6d2f636c612d70726f6a6563742d6c6f676f2d70726f642f636c612d6d697373696e672d69642e737667">](https://api.easycla.lfx.linuxfoundation.org/v2/repository-provider/github/sign/29311126/65600975/107910/#/?version=2) [<img alt="CLA Not Signed" width="328" height="28" src="https://camo.githubusercontent.com/386c32cf1de823184a8b4907a7c1cedf6e9c4f628cdb9e0007df4a36420e9b87/68747470733a2f2f73332e616d617a6f6e6177732e636f6d2f636c612d70726f6a6563742d6c6f676f2d70726f642f636c612d6e6f742d7369676e65642e737667">](https://api.easycla.lfx.linuxfoundation.org/v2/repository-provider/github/sign/29311126/65600975/107910/#/?version=2)
> 
> * ✅ login: voznesenskym / name: Michael Voznesensky  ([7aff928](https://github.com/pytorch/pytorch/commit/7aff9284394f5884ef15b6aed8f175012e4ec4b2))
> * ❌ The email address for the commit ([6498a82](https://github.com/pytorch/pytorch/commit/6498a8223f5084e3454e9fd963ca06a021aefb17), [9cd45a6](https://github.com/pytorch/pytorch/commit/9cd45a659cba96a134ffcc04e89b09c78eb48b87)) is not linked to the GitHub account, preventing the EasyCLA check. Consult [this Help Article](https://confluence.linuxfoundation.org/pages/viewpage.action?pageId=86641302) and [GitHub Help](https://help.github.com/en/github/committing-changes-to-your-project/why-are-my-commits-linked-to-the-wrong-user) to resolve. (To view the commit's email address, add .patch at the end of this PR page's URL.) For further assistance with EasyCLA, [please submit a support request ticket](https://jira.linuxfoundation.org/servicedesk/customer/portal/4).

fixed

mlazos committed:
Add tests for foreach copy (#107860)

Fixes #ISSUE_NUMBER

Pull Request resolved: https://github.com/pytorch/pytorch/pull/107860
Approved by: https://github.com/eellison, https://github.com/jansel

gmagogsfm committed:
Make validate private by renaming validate to _validate (#107927)

Pull Request resolved: https://github.com/pytorch/pytorch/pull/107927
Approved by: https://github.com/tugsbayasgalan

pearu committed:
Support torch.sparse_mask on strided input with sparse CSR, CSC, BSR, and BSC mask. (#107777)

While `input.sparse_mask(mask)` can be defined as `input.mul(ones_like(mask))`, implementing this definition leads to a chicken-and-egg problem because the multiplication of dense and sparse_compressed tensors relies on the `sparse_mask` support.

This PR implements `sparse_mask` support for sparse compressed masks using utility functions from sparse compressed tensor conversions support.

Fixes https://github.com/pytorch/pytorch/issues/107373
Fixes https://github.com/pytorch/pytorch/issues/107370

Pull Request resolved: https://github.com/pytorch/pytorch/pull/107777
Approved by: https://github.com/amjames, https://github.com/cpuhrsch
ghstack dependencies: #107638

pearu committed:
Add torch.sparse.as_sparse_gradcheck decorator of gradcheck that allows gradcheck input function to receive and return sparse tensors (#107150)

Compared to #104848, this PR makes a step further: when the enable_sparse_support decorator is applied to `torch.autograd.gradcheck`, the resulting callable is equivalent to `torch.autograd.gradcheck` with an extra feature of supporting functions that can have input sparse tensors or/and can return sparse tensors.

At the same time, the underlying call to `torch.autograd.gradcheck` will operate on strided tensors only. This basically means that torch/autograd/gradcheck.py can be cleaned up by removing the code that deals with sparse tensors.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/107150
Approved by: https://github.com/albanD, https://github.com/amjames, https://github.com/cpuhrsch
ghstack dependencies: #107638, #107777

jataylo committed:
[ROCm] Use rocm manylinux builder image for triton wheels (#107600)

Update to ROCm triton pinned commit for the 2.1 branch cut off.

As part of this we are updating `build_triton_wheel.py` and `build-triton-wheel.yml` to support building ROCm triton wheels through pytorch/manylinux-rocm to avoid the need of slowly downloading rpm libraries for ROCm in the cpu manylinux builder image and avoiding the need to maintain a conditional file with hard coded repositories from radeon.org for every ROCm release.

This new approach will allow us to build wheels faster in a more easily maintainable way.

This PR also brings in a required change as Triton on ROCm requires device_type to be set to hip so we can pass down the correct device type to triton (https://github.com/ROCmSoftwarePlatform/triton/pull/284).

Pull Request resolved: https://github.com/pytorch/pytorch/pull/107600
Approved by: https://github.com/jansel, https://github.com/jithunnair-amd

dependabot[bot] committed:
Bump scipy from 1.9.3 to 1.10.1 in /.ci/docker (#104746)

* Bump scipy from 1.9.3 to 1.10.1 in /.ci/docker

Bumps [scipy](https://github.com/scipy/scipy) from 1.8.1 to 1.10.0.
- [Release notes](https://github.com/scipy/scipy/releases)
- [Commits](https://github.com/scipy/scipy/compare/v1.8.1...v1.10.0)

---
updated-dependencies:
- dependency-name: scipy
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>

* Update .ci/docker/requirements-ci.txt

---------

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Nikita Shulga <nshulga@meta.com>

aakhundov committed:
[inductor] Add cat + split_with_sizes elimination pass (#107956)

Summary:

When the `cat` inputs' sizes and the `split_sizes` of the downstream `split_with_sizes` match, the `cat` + `split_with_sizes` constellation can be eliminated. E.g. here:

```
@torch.compile
def fn(a, b, c):
    cat = torch.ops.aten.cat.default([a, b, c], 1)
    split_with_sizes = torch.ops.aten.split_with_sizes.default(cat, [2, 3, 5], 1)
    return [s ** 2 for s in split_with_sizes]

inputs = [
    torch.randn(2, 2, device="cuda"),
    torch.randn(2, 3, device="cuda"),
    torch.randn(2, 5, device="cuda"),
]
output = fn(*inputs)
```

This PR adds a new fx pass for such elimination. The new pass is similar to the existing [`splitwithsizes_cat_replace`](https://github.com/pytorch/pytorch/blob/b18e1b684a7673daa3a51128aae4e75ed7aa7cbc/torch/_inductor/fx_passes/post_grad.py#L508), but considers the ops in the opposite order.

Test Plan:

```
$ python test/inductor/test_pattern_matcher.py

...

----------------------------------------------------------------------
Ran 21 tests in 46.450s

OK
```

Reviewers:

Subscribers:

Tasks:

Tags:

Pull Request resolved: https://github.com/pytorch/pytorch/pull/107956
Approved by: https://github.com/jansel

dilililiwhy committed:
Enable custom device support in fsdp checkpoint (#107289)

Fixes https://github.com/pytorch/pytorch/issues/104390
Enable custom device(privateuse1 backend) support in checkpointing by a dynamic abstract device module.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/107289
Approved by: https://github.com/wz337

Xia-Weiwen committed:
[ONEDNN][BC-breaking] update onednn from v2.7.3 to v3.1.1 (#97957)

**Summary**
Update onednn from v2.7.3 to v3.1.1.
It is bc-breaking as some APIs are changed on oneDNN side. Changes include:
- PyTorch code where oneDNN is directly called
- Submodule `third_party/ideep` to adapt to oneDNN's new API.
- CMAKE files to fix build issues.

**Test plan**
Building issues and correctness are covered by CI checks.
For performance, we have run TorchBench models to ensure there is no regression. Below is the comparison before and after oneDNN update.
![image](https://github.com/pytorch/pytorch/assets/12522207/415a4ff0-7566-40c6-aed0-24997a475b0e)

Note:
- Base commit of PyTorch: da322ea
- CPU: Intel(R) Xeon(R) Platinum 8380 CPU @ 2.30GHz (Ice Lake)

Pull Request resolved: https://github.com/pytorch/pytorch/pull/97957
Approved by: https://github.com/jgong5, https://github.com/jerryzh168

leslie-fang-intel committed:
[Quant][PT2E] Enable qconv for quantization 2.0 export (#104580)

**Summary**
Enable `qconv1d/2d/3d`, `qconv2d_relu`, `qconv2d_add`, and `qconv2d_add_relu` operator for quantization 2.0 export with oneDNN library.

**Test Plan**
```
python -u -m pytest -s -v test_quantized_op.py -k test_qconv1d_pt2e
python -u -m pytest -s -v test_quantized_op.py -k test_qconv2d_pt2e
python -u -m pytest -s -v test_quantized_op.py -k test_qconv3d_pt2e
python -u -m pytest -s -v test_quantized_op.py -k test_qconv2d_relu_pt2e
python -u -m pytest -s -v test_quantized_op.py -k test_qconv2d_add_pt2e
python -u -m pytest -s -v test_quantized_op.py -k test_qconv2d_add_relu_pt2e
```

Pull Request resolved: https://github.com/pytorch/pytorch/pull/104580
Approved by: https://github.com/jgong5, https://github.com/jerryzh168

leslie-fang-intel committed:
[Quant][Inductor] Enable quantized conv weight prepack inside inductor constant folding (#104581)

**Summary**
Enable quantization conv weight prepack inside inductor constant folding.

**Test Plan**
```
python -m pytest test_mkldnn_pattern_matcher.py -k test_qconv2d_unary
```

Pull Request resolved: https://github.com/pytorch/pytorch/pull/104581
Approved by: https://github.com/jgong5, https://github.com/eellison
ghstack dependencies: #104580

leslie-fang-intel committed:
[Quant][Inductor] Enable quantization conv pattern fusion inside inductor (#104588)

**Summary**
Enable the `dequant-quantization-quant` pattern fusion and lowering inside inductor.

**Test Plan**
```
python -m pytest test_mkldnn_pattern_matcher.py -k test_qconv2d_unary
```

Pull Request resolved: https://github.com/pytorch/pytorch/pull/104588
Approved by: https://github.com/jgong5, https://github.com/eellison
ghstack dependencies: #104580, #104581

leslie-fang-intel committed:
[Quant][Inductor] Enable dequant promotion inside inductor (#104590)

**Summary**
Enable the `dequant pattern` promotion pass in inductor. Since in the qconv weight prepack pass, we will match the `dequant->conv2d` pattern. If the `dequant pattern` has multi user nodes, it will fail to be matched.
Taking the example of
```
        conv1
       /     \
   conv2    conv3
```
After quantization flow, it will generate pattern as
```
      dequant1
          |
        conv1
          |
        quant2
          |
       dequant2
       /     \
   conv2    conv3
```
We need to duplicate `dequant2` into `dequant2` and `dequant3`, in order to make `dequant2->conv2` and  `dequant3->conv3`  pattern matched.

**Test Plan**
```
python -m pytest test_mkldnn_pattern_matcher.py -k test_dequant_promotion
```

Pull Request resolved: https://github.com/pytorch/pytorch/pull/104590
Approved by: https://github.com/jgong5, https://github.com/eellison
ghstack dependencies: #104580, #104581, #104588

leslie-fang-intel committed:
[Quant][Inductor] Enable quantization conv_unary(relu) pattern fusion inside inductor (#105455)

**Summary**
Enable the `dequant-conv2d-unary_postop(relu)-quant` pattern fusion and lowering inside inductor.

**Test Plan**
```
clear && python -m pytest test_mkldnn_pattern_matcher.py -k test_qconv2d_unary
```

Pull Request resolved: https://github.com/pytorch/pytorch/pull/105455
Approved by: https://github.com/jgong5, https://github.com/eellison
ghstack dependencies: #104580, #104581, #104588, #104590

leslie-fang-intel committed:
[Quant][Inductor] Enable quantization conv_binary(add/add_relu) pattern fusion inside inductor (#105456)

**Summary**
Enable the `dequant-conv2d-binary_postop(add)-unary_postop(relu)-quant` pattern fusion and lowering inside inductor.

**Test Plan**
```
clear && python -m pytest test_mkldnn_pattern_matcher.py -k test_qconv2d_binary
```

Pull Request resolved: https://github.com/pytorch/pytorch/pull/105456
Approved by: https://github.com/jgong5, https://github.com/eellison
ghstack dependencies: #104580, #104581, #104588, #104590, #105455

leslie-fang-intel committed:
[Quant][PT2E] Enable X86InductorQuantizer single quantizable op(maxpool2d) (#105639)

**Summary**
In this PR, we mainly enable 2 things.

- Enable the skeleton of quantization recipe for single quantizable operators in `X86InductorQuantizer`.
- Add quantization recipe of `maxpool2d` and annotate it as input./output share observer.

**Test Plan**
```
python -m pytest test_x86inductor_quantizer.py -k test_maxpool2d_recipe
```

Pull Request resolved: https://github.com/pytorch/pytorch/pull/105639
Approved by: https://github.com/jgong5, https://github.com/jerryzh168
ghstack dependencies: #104580, #104581, #104588, #104590, #105455, #105456

albanD committed:
Initial Python 3.12 build fixes (#106083)

This compiles with python 3.12
You can get numpy from https://anaconda.org/scientific-python-nightly-wheels/numpy/files so that you don't need to remove numpy from test files.

Basic core tests work but obviously dynamo and first class dims don't work.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/106083
Approved by: https://github.com/ezyang

pytorchmergebot committed:
Revert "[optim] Make casting to match params a hook (#106725)"

This reverts commit 9f86d8517201a3d473a8e80d12e22b46570c88a2.

Reverted https://github.com/pytorch/pytorch/pull/106725 on behalf of https://github.com/janeyx99 due to We acknowledge this is a huge risk because people do not remember to call super().__init__ from their Optimizer subclasses and so this will break lots of load_state_dict behavior ([comment](https://github.com/pytorch/pytorch/pull/106725#issuecomment-1693386137))

DanilBaibak committed:
Cleanup RUNNER_TEMP folder (#107868)

Cleanup RUNNER_TEMP folder
Pull Request resolved: https://github.com/pytorch/pytorch/pull/107868
Approved by: https://github.com/atalman

jerryzh168 committed:
[quant][pt2e] Enable testing for reference quant model representations (#107474)

Summary:
Previously these tests were disabled due to time out in dynamo export in fbcode,
this might have been resolved, so trying to enable again

Test Plan:
python test/test_quantization.py TestQuantizePT2E

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D48619072](https://our.internmc.facebook.com/intern/diff/D48619072)
Pull Request resolved: https://github.com/pytorch/pytorch/pull/107474
Approved by: https://github.com/andrewor14

angelayi committed:
[core aten] Remove split.Tensor from core aten (#107938)

Removing split.Tensor from core aten as it can be decomposed into split_with_sizes

Pull Request resolved: https://github.com/pytorch/pytorch/pull/107938
Approved by: https://github.com/larryliu0820

hongxiayang committed:
[ROCm] enable missed cpp tests - test_libtorch_jit (test_jit and test_lazy) (#107234)

[ROCm] enable missed cpp tests - test_libtorch_jit (test_jit and test_lazy)

Pull Request resolved: https://github.com/pytorch/pytorch/pull/107234
Approved by: https://github.com/jithunnair-amd, https://github.com/pruthvistony, https://github.com/ezyang

angelayi committed:
[export] Store the arguments used to trace the exported program in itself (#107906)

Proper fix would be to do something like https://github.com/pytorch/pytorch/pull/107877, but since that depends on internal changes and it would take too long for diff train to land we will first just make OSS work using torch.save.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/107906
Approved by: https://github.com/gmagogsfm

pytorchmergebot committed:
Revert "[Dynamo] cache_size policy (#107496)"

This reverts commit 4175a6e9443c58d57e0fc595f506c43aec8cb477.

Reverted https://github.com/pytorch/pytorch/pull/107496 on behalf of https://github.com/ZainRizvi due to Breaking internal builds ([comment](https://github.com/pytorch/pytorch/pull/107496#issuecomment-1693590121))

Anstow committed:
Added normal op decomposition for specializations of the normal op (#106792)

This fixes running normal with the meta key.

```
import torch

t = torch.tensor(4.0, device='meta')
torch.normal(0.5, t)
```

Pull Request resolved: https://github.com/pytorch/pytorch/pull/106792
Approved by: https://github.com/lezcano

gmagogsfm committed:
Hide `transform` method by renaming it (#107940)

Pull Request resolved: https://github.com/pytorch/pytorch/pull/107940
Approved by: https://github.com/tugsbayasgalan

janeyx99 committed:
Fixes #107737 SGD doc blank line (#107738)

docs preview brings joy
<img width="774" alt="image" src="https://github.com/pytorch/pytorch/assets/31798555/1bfaae64-16f2-448a-8af2-36303d2845db">

Pull Request resolved: https://github.com/pytorch/pytorch/pull/107738
Approved by: https://github.com/mikaylagawarecki

wconstab commented on test/dynamo/test_misc.py:
```diff
@@ -6750,6 +6750,82 @@ def main_generator(t_list):
+
+    def test_itertools_accumulate_tensors_builtins(self):
+        for builtin_op in [operator.mul, operator.sub, operator.pow]:
+            torch._dynamo.reset()
```
why do you need a reset here? shouldn't we properly cache miss on new builtin op?

voznesenskym replied:
Yup! Don't need it!

wconstab replied:
oh, ok, great.

well there is one more reset to remove but ill stamp.

wconstab commented on torch/_dynamo/variables/misc.py:
```diff
@@ -714,6 +714,43 @@ def call_function(
+                        acc = func(tx, item, acc)
+                    except:
+                        raise unimplemented(
+                            "Unexpected failure in invoking function during accumulate"
```
would it help to include the exception msg? presumably in practice if you hit this it'd be some type mismatch or something useful to know

xwang233 committed:
[cuda] Fix the incorrect types in int8_gemm (#107895)

Fixes #107671

From cublas team: alpha and beta need to be of the same C++ type as of scaleType, which is `int32_t` here.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/107895
Approved by: https://github.com/Skylion007, https://github.com/cpuhrsch

tugsbayasgalan committed:
Support cond and out_dtype for predispatch (#107941)

Summary: Title

Test Plan: CI

Differential Revision: D48675742

Pull Request resolved: https://github.com/pytorch/pytorch/pull/107941
Approved by: https://github.com/jerryzh168

Anstow committed:
Implement size checking for copy_ with meta tensors (#107779)

Pull Request resolved: https://github.com/pytorch/pytorch/pull/107779
Approved by: https://github.com/ezyang

CaoE committed:
Add backward check for test_memory_format (#106104)

Add backward check for test_memory_format.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/106104
Approved by: https://github.com/mikaylagawarecki

jansel committed:
[dynamo] Treat monkey patched .forward as dynamic (#107104)

Pull Request resolved: https://github.com/pytorch/pytorch/pull/107104
Approved by: https://github.com/anijain2305

pytorchmergebot committed:
Revert "Remove remaining global `set_default_dtype` calls from tests (#107246)"

This reverts commit aa8ea1d787a9d21b064b664c5344376265feea6c.

Reverted https://github.com/pytorch/pytorch/pull/107246 on behalf of https://github.com/facebook-github-bot due to Diff reverted internally ([comment](https://github.com/pytorch/pytorch/pull/107246#issuecomment-1693838522))

fduwjj committed:
[TP][2D][EZ] Fix Error in FSDP 2D test (#107975)

As title, TP dimension should be the second dim, so we need to pass tp_degree to the second rather the first dim of the mesh tensor.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/107975
Approved by: https://github.com/wz337, https://github.com/awgu

eellison commented on torch/_dynamo/variables/misc.py:
```diff
@@ -714,6 +714,45 @@ def call_function(
             return variables.ListIteratorVariable(
                 items, mutable_local=MutableLocal(), **options
             )
+        elif self.value is itertools.accumulate and not kwargs:
```
would be nice if we could just register pure-python accumulate implementations:

```
def accumulate(iterable, func=lambda x, y: x + y):
    it = iter(iterable)
    try:
        # Initialize the accumulator with the first value from the iterable
        accumulator = next(it)
    except StopIteration:
        # If the iterable is empty, return an empty generator
        return
    yield accumulator
    
    # Continue accumulating using func and yield each accumulated value
    for element in it:
        accumulator = func(accumulator, element)
        yield accumulator
```

(from chat gpt)

voznesenskym replied:
@eellison why wouldn't this work after https://github.com/pytorch/pytorch/pull/106986/files?

```
    def test_pure_python_accumulate(self):
        def accumulate(iterable, func=lambda x, y: x + y):
            it = iter(iterable)
            try:
                # Initialize the accumulator with the first value from the iterable
                accumulator = next(it)
            except StopIteration:
                # If the iterable is empty, return an empty generator
                return
            yield accumulator

            for element in it:
                accumulator = func(accumulator, element)
                yield accumulator

        def fn(it):
            return accumulate(it)

        t_list = [torch.tensor([i]) for i in range(4)]
        eager = fn(t_list)

        counter = CompileCounter()
        compiled_fn = torch._dynamo.optimize(counter)(fn)
        compiled = compiled_fn(t_list)

        self.assertEqual(list(eager), list(compiled))
        self.assertEqual(counter.frame_count, 1)
```

This test passes

eellison replied:
cool ! could we use the pure python lowering to support accumulate ? instead of writing dynamo, write python (kind of like a decomposition)

voznesenskym replied:
Hmm, I suppose? I don't know what the value of that is though - seems more brittle?

andrewor14 committed:
[quant][pt2e] Fix annotation for conv no bias case (#107971)

Summary: This fixes the no bias case for conv annotations.
Previously this would result in an index out of bounds, since
the new aten.conv2d op may not have the bias arg (unlike the
old aten.convolution op). This was not caught because of a lack
of test cases, which are added in this commit.

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_qat_conv_no_bias
python test/test_quantization.py TestQuantizePT2E.test_qat_conv_bn_relu_fusion_no_conv_bias

Reviewers: jerryzh168, kimishpatel

Subscribers: jerryzh168, kimishpatel

Differential Revision: [D48696874](https://our.internmc.facebook.com/intern/diff/D48696874)
Pull Request resolved: https://github.com/pytorch/pytorch/pull/107971
Approved by: https://github.com/jerryzh168

CaoE committed:
add channel last 3d support for maxpool3d on CPU (#97775)

### Testing
Single socket (28 cores):

shape | fp32 forward / ms | bf16 forward / ms | fp32 backward / ms  | bf16 backward / ms
-- | -- | -- | -- | --
size: (1, 56, 264, 264), kernel: 3,   stride: 1, mem_format: contig | 3.959584 | 5.493402 | 0.557232 | 0.568485
size: (1, 56, 264, 264), kernel: 3,   stride: 1, mem_format: CL | 0.815511 | 1.351261 | 5.710506 | 10.57506
size: (32, 32, 100, 100), kernel: 3,   stride: 1, mem_format: contig  | 10.63426 | 15.28637 | 2.67656 | 1.71365
size: (32, 32, 100, 100), kernel: 3,   stride: 1, mem_format: CL | 2.63570 | 2.05532 | 2.55452 | 2.33923
size: (4, 19, 10, 16, 16), kernel:   3, stride: 1, mem_format: contig | 0.375469 | 0.479748 | 0.066364 | 0.065155
size: (4, 19, 10, 16, 16), kernel:   3, stride: 1, mem_format: CL3d | 0.112197 | 0.112326 | 0.111697 | 0.145364

Single core:

shape | fp32 forward / ms | bf16 forward / ms | fp32 backward / ms | bf16 backward / ms
-- | -- | -- | -- | --
size: (1, 56, 264, 264), kernel: 3,   stride: 1, mem_format: contig | 92.16582 | 128.6513 | 6.684325 | 12.21541
size: (1, 56, 264, 264), kernel: 3,   stride: 1, mem_format: CL | 10.14318 | 29.80297 | 7.350142 | 11.25323
size: (32, 32, 100, 100), kernel: 3,   stride: 1, mem_format: contig | 238.55453 | 331.89967 | 19.694657 | 32.78853
size: (32, 32, 100, 100), kernel: 3,   stride: 1, mem_format: CL | 30.17079 | 32.75628 | 22.44543 | 30.17796
size: (4, 19, 10, 16, 16), kernel:   3, stride: 1, mem_format: contig | 7.474389 | 9.937217 | 0.236015 | 0.434229
size: (4, 19, 10, 16, 16), kernel:   3, stride: 1, mem_format: CL3d | 2.318954 | 2.469444 | 0.262125 | 0.401361

Pull Request resolved: https://github.com/pytorch/pytorch/pull/97775
Approved by: https://github.com/jgong5, https://github.com/mikaylagawarecki

cyyever committed:
[BE] use DeviceIndex instead of int64_t for related device interfaces (#103068)

This PR unifies the device interfaces in aten/*cpp and torch/csrc/*cpp to use  **c10::DeviceIndex**.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/103068
Approved by: https://github.com/malfet

mlazos committed:
Handle 2D blocking with foreach (#107840)

Previously blocking in foreach ops was only 1D. This PR allows handling kernels with 2D blocking with foreach as well.

Code when at least one dim matches:
[example code + output](https://gist.github.com/mlazos/9f100b21cfe2540f0a24303a8349c196)

Code when neither X or Y dim matches:
[example code + output](https://gist.github.com/mlazos/14e2a455f635896dface09be601595dd)

Pull Request resolved: https://github.com/pytorch/pytorch/pull/107840
Approved by: https://github.com/jansel

shunting314 committed:
[reland][inductor] make thread order consistent with loop order (#107902)

This PR relands https://github.com/pytorch/pytorch/pull/106827 which get reverted because of causing compilation error for some ads model.

Yanbo provide a repro in one of the 14k model ( `pytest ./generated/test_KaiyangZhou_deep_person_reid.py -k test_044`). This is also the model I used to confirm the fix and come up with a unit test. In this model, we call `tritoin_heuristics.triton_config` with size_hints [2048, 2]. Previously this would result in a trition config with XBLOCK=2048 and YBLOCK=2 . But since we change the mapping between size_hints and XYZ dimension, we now generate a triton config with XBLOCK=2 and YBLOCK=2048.  This fails compilation since we set max YBLOCK to be 1024.

My fix is to make sure we never generate a triton config that exceeds the maximum block size.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/107902
Approved by: https://github.com/jansel

hl475 committed:
AOTInductor: error: ‘c10::Dispatcher’ has not been declared for CPU model (#107935)

Differential Revision: D48675055

Pull Request resolved: https://github.com/pytorch/pytorch/pull/107935
Approved by: https://github.com/houseroad

titaiwangms committed:
[ONNX] Support constant tensors in FakeMode exporting (#107836)

Fixes https://github.com/pytorch/pytorch/issues/107475

- Constant tensors was wrongly recognized as weights and buffers, and then was detached from its default value during `to_model_proto`. This PR fixes the bug and pick up Bloom CI test back successfully. NOTE: non-persistent buffer and weights has different situation and is not fixed by this PR.
- Reduce transformers model size by modifying their config parameters to speed up CI tests. (Unrelated to this PR title)

Corresponding change with https://github.com/microsoft/onnxscript/pull/1023
Pull Request resolved: https://github.com/pytorch/pytorch/pull/107836
Approved by: https://github.com/BowenBao, https://github.com/justinchuby

BowenBao committed:
[ONNX] Remove API reference for TorchScript export diagnostics (#107979)

Remove both api reference and rules specific to TorchScript ONNX export. The page should display only info related to `torch.onnx.dynamo_export` diagnostics.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/107979
Approved by: https://github.com/justinchuby

JackCaoG committed:
Make openxla and opexla_eval backend show up in list_backends (#107905)

The reason to keep the non-aot(openxla_eval) backend is discussed in https://github.com/pytorch/xla/issues/5430#issuecomment-1683191662.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/107905
Approved by: https://github.com/jansel

tugsbayasgalan committed:
Add runtime assertions for prim values (#107939)

Pull Request resolved: https://github.com/pytorch/pytorch/pull/107939
Approved by: https://github.com/gmagogsfm

Commit:
[RFC] Change --standalone to bind to a random port (#107734)

Given standalone generates args anyways, it seems like it would be more convenient if it explicitly used a random port by default instead of trying to use 29400.

That way users can directly go with `--standalone` instead of having to spell out `--rdzv-backend=c10d --rdzv-endpoint=localhost:0`

Pull Request resolved: https://github.com/pytorch/pytorch/pull/107734
Approved by: https://github.com/H-Huang

pytorchmergebot committed:
Revert "Handle 2D blocking with foreach (#107840)"

This reverts commit f87ffe473d4825c15eaea0360baf08cad49979de.

Reverted https://github.com/pytorch/pytorch/pull/107840 on behalf of https://github.com/huydhn due to Sorry for reverting this, but test_2d_blocking is failing in trunk, probably a landrace as PR was green ([comment](https://github.com/pytorch/pytorch/pull/107840#issuecomment-1694009217))

tugsbayasgalan committed:
Disable Constraint constructor (#107918)

Pull Request resolved: https://github.com/pytorch/pytorch/pull/107918
Approved by: https://github.com/zhxchen17

jerryzh168 committed:
[quant][pt2e][xnnpack_quantizer] Add support for mul and mul_relu (#107930)

Summary: att

Test Plan: buck2 run executorch/examples/quantization:example -- -m=mv3 --verify

Differential Revision: D48588121

Pull Request resolved: https://github.com/pytorch/pytorch/pull/107930
Approved by: https://github.com/kimishpatel

leslie-fang-intel committed:
[Quant][Inductor] Enable the lowering of quantized maxpool2d (#105906)

**Summary**
Enable the `dq-maxpool2d-q` pattern match and lower into `torch.ops.quantized.max_pool2d`.

**Test Plan**
```
python -m pytest test_mkldnn_pattern_matcher.py -k test_qmaxpool2d
python -m pytest test_quantized_op.py -k test_max_pool2d_pt2e
```

Pull Request resolved: https://github.com/pytorch/pytorch/pull/105906
Approved by: https://github.com/jgong5, https://github.com/eellison
ghstack dependencies: #104580, #104581, #104588, #104590, #105455, #105456, #105639

leslie-fang-intel committed:
[Quant][PT2E] Enable weight scale optimization in QConv PT2E (#105996)

**Summary**
After oneDNN 3.1 upgrade, we don't need to do the weight scale reciprocal calculation. So, remove the redundant reciprocal calculation to optimize QConv performance and using IDeep version API to implement it in this PR:

- This QConv implementation expects to work functionally both with current IDeep version and the following IDeep upgrade in PR: https://github.com/pytorch/pytorch/pull/107565.
- With the following IDeep upgrade in PR: https://github.com/pytorch/pytorch/pull/107565, the QConv has better performance since the redundant reciprocal calculation are removed.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/105996
Approved by: https://github.com/jgong5, https://github.com/jerryzh168
ghstack dependencies: #104580, #104581, #104588, #104590, #105455, #105456, #105639, #105906

leslie-fang-intel committed:
[submodule][Quant][PT2E] Upgrade IDeep to remove redundant QConv weight scale reciprocal calculation (#107565)

**Summary**
Upgrade IDeep which includes 1 IDeep change as IDeep PR: https://github.com/intel/ideep/pull/226

- For IDeep PR: https://github.com/intel/ideep/pull/226 which has done 2 things:

  - Remove the redundant QConv weight scale reciprocal calculation.
  - Pump IDEEP_VERSION_REVISION version from 0 to 1.

  So only QConv related calculation will be impacted and we already use IDeep version API in https://github.com/pytorch/pytorch/pull/105996 to make the corresponding change in PyTorch.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/107565
Approved by: https://github.com/jgong5, https://github.com/jerryzh168
ghstack dependencies: #104580, #104581, #104588, #104590, #105455, #105456, #105639, #105906, #105996

voznesenskym commented:
@pytorchbot merge

voznesenskym commented:
@pytorchbot merge

drisspg committed:
Bias overflow fix mem eff bias (#107968)

Fixes #107959
This should have been fixed here https://github.com/pytorch/pytorch/pull/103201
Edit:
Looking at git blame it appears the dropout revet squashed the changes from this PR
Pull Request resolved: https://github.com/pytorch/pytorch/pull/107968
Approved by: https://github.com/cpuhrsch

pytorchmergebot committed:
Revert "[quant][pt2e][xnnpack_quantizer] Add support for mul and mul_relu (#107930)"

This reverts commit 1d1739dc6d7365c28719cd0175081f9d9aab0324.

Reverted https://github.com/pytorch/pytorch/pull/107930 on behalf of https://github.com/facebook-github-bot due to Diff reverted internally ([comment](https://github.com/pytorch/pytorch/pull/107930#issuecomment-1694069330))

angelayi committed:
[export] Serialize getattr nodes (#107924)

Turns out some graphs will result in getattr nodes...so let's serialize them
Pull Request resolved: https://github.com/pytorch/pytorch/pull/107924
Approved by: https://github.com/zhxchen17, https://github.com/avikchaudhuri

JackCaoG committed:
Minor fixs to make torchbench runable on torch/xla (#107919)

`import torch_xla.core.xla_model as xm` no longer trigger the xla runtime to init, hence explictly create the device here. This is a workaround for https://github.com/pytorch/xla/issues/4174.

`is_correct` reference has been deleted, I think it is a deadcode.

After this patch, I am able to run

```
python benchmarks/dynamo/torchbench.py --randomize-input --performance --trace-on-xla --training --backend=openxla --only resnet50
```

Pull Request resolved: https://github.com/pytorch/pytorch/pull/107919
Approved by: https://github.com/shunting314, https://github.com/wconstab

CYuxian committed:
[ONNX] Return input itself for non-fp inputs and support decimals for aten::round op (#107920)

Pull Request resolved: https://github.com/pytorch/pytorch/pull/107920
Approved by: https://github.com/justinchuby

wz337 committed:
[DeviceMesh] Expose init_device_mesh (#107969)

As title.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/107969
Approved by: https://github.com/wanchaol, https://github.com/fduwjj

leslie-fang-intel committed:
[Quant][PT2E] Add cat and avg_pool2d recipe into x86InductorQuantizer (#106836)

**Summary**
Add `cat` and `avg_pool2d` quantization recipe as input output share observer into `x86InductorQuantizer`.

**Test Plan**
```
clear && python -m pytest test_x86inductor_quantizer.py -k test_cat_recipe
clear && python -m pytest test_x86inductor_quantizer.py -k test_cat_recipe_same_inputs
clear && python -m pytest test_x86inductor_quantizer.py -k test_cat_recipe_single_input
clear && python -m pytest test_x86inductor_quantizer.py -k test_avg_pool2d_recipe
```

Pull Request resolved: https://github.com/pytorch/pytorch/pull/106836
Approved by: https://github.com/jgong5, https://github.com/jerryzh168

leslie-fang-intel committed:
[Quant][Inductor] Enable lowering of qcat (#106838)

**Summary**
Enable the lowering of `qcat` inside inductor as extern kernel.

**Test Plan**
```
python -m pytest test_mkldnn_pattern_matcher.py -k test_qcat
```

Pull Request resolved: https://github.com/pytorch/pytorch/pull/106838
Approved by: https://github.com/jgong5, https://github.com/eellison
ghstack dependencies: #106836

leslie-fang-intel committed:
[Quant][Inductor] Fix the non contiguous load with uint8 data type (#106958)

**Summary**
Currently, the load vectorization code generation with `non_contiguous` and `uint8` data type has issue in determining the data type. It caused wrong results in `shufflenet_v2_x1_0` model after we enable the `cat` quantization recipe.

- Previously code gen with the example in this PR:

```
cpp_fused_clone_view_0 = async_compile.cpp('''
#include "/tmp/torchinductor_root/i5/ci5uspp363v3ky6jkccllm3bxudy2fkdpqinkqhmpehfihejs7ko.h"
extern "C" void kernel(const unsigned char* in_ptr0,
                       float* out_ptr0)
{
    #pragma omp parallel num_threads(56)
    {
        {
            #pragma omp for
            for(long i0=static_cast<long>(0L); i0<static_cast<long>(232L); i0+=static_cast<long>(1L))
            {
                for(long i1=static_cast<long>(0L); i1<static_cast<long>(784L); i1+=static_cast<long>(16L))
                {
                    auto tmp0 = ([&]() { __at_align__ float tmpbuf[16]; for (long i1_inner = 0; i1_inner < 16; i1_inner++) tmpbuf[i1_inner] = flag_to_float_scalar(in_ptr0[static_cast<long>((116L*(static_cast<long>(i0) % static_cast<long>(2L))) + (232L*i1) + (232L*i1_inner) + (at::native::div_floor_integer(i0, 2L)))]); return at::vec::Vectorized<uint8_t>::loadu_one_fourth(tmpbuf); })();
                    auto tmp1 = at::vec::convert_uint8_to_float(tmp0);
                    auto tmp2 = at::vec::Vectorized<float>(static_cast<float>(0.0));
                    auto tmp3 = tmp1 - tmp2;
                    auto tmp4 = at::vec::Vectorized<float>(static_cast<float>(1.0));
                    auto tmp5 = tmp3 * tmp4;
                    auto tmp6 = tmp5 * tmp4;
                    auto tmp7 = tmp6.round();
                    auto tmp8 = tmp7 + tmp2;
                    auto tmp9 = at::vec::maximum(tmp8, tmp2);
                    auto tmp10 = at::vec::Vectorized<float>(static_cast<float>(255.0));
                    auto tmp11 = at::vec::minimum(tmp9, tmp10);
                    auto tmp12 = at::vec::convert_float_to_uint8(tmp11);
                    auto tmp13 = at::vec::convert_uint8_to_float(tmp12);
                    auto tmp14 = tmp13 - tmp2;
                    auto tmp15 = tmp14 * tmp4;
                    tmp15.store(out_ptr0 + static_cast<long>(i1 + (784L*i0)));
                }
            }
        }
    }
}
''')
```

- After this PR, the code gen is:

```
cpp_fused_clone_view_0 = async_compile.cpp('''
#include "/tmp/torchinductor_root/i5/ci5uspp363v3ky6jkccllm3bxudy2fkdpqinkqhmpehfihejs7ko.h"
extern "C" void kernel(const unsigned char* in_ptr0,
                       float* out_ptr0)
{
    #pragma omp parallel num_threads(56)
    {
        {
            #pragma omp for
            for(long i0=static_cast<long>(0L); i0<static_cast<long>(232L); i0+=static_cast<long>(1L))
            {
                for(long i1=static_cast<long>(0L); i1<static_cast<long>(784L); i1+=static_cast<long>(16L))
                {
                    auto tmp0 = ([&]() { __at_align__ unsigned char tmpbuf[16]; for (long i1_inner = 0; i1_inner < 16; i1_inner++) tmpbuf[i1_inner] = in_ptr0[static_cast<long>((116L*(static_cast<long>(i0) % static_cast<long>(2L))) + (232L*i1) + (232L*i1_inner) + (at::native::div_floor_integer(i0, 2L)))]; return at::vec::Vectorized<uint8_t>::loadu_one_fourth(tmpbuf); })();
                    auto tmp1 = at::vec::convert_uint8_to_float(tmp0);
                    auto tmp2 = at::vec::Vectorized<float>(static_cast<float>(0.0));
                    auto tmp3 = tmp1 - tmp2;
                    auto tmp4 = at::vec::Vectorized<float>(static_cast<float>(1.0));
                    auto tmp5 = tmp3 * tmp4;
                    auto tmp6 = tmp5 * tmp4;
                    auto tmp7 = tmp6.round();
                    auto tmp8 = tmp7 + tmp2;
                    auto tmp9 = at::vec::maximum(tmp8, tmp2);
                    auto tmp10 = at::vec::Vectorized<float>(static_cast<float>(255.0));
                    auto tmp11 = at::vec::minimum(tmp9, tmp10);
                    auto tmp12 = at::vec::convert_float_to_uint8(tmp11);
                    auto tmp13 = at::vec::convert_uint8_to_float(tmp12);
                    auto tmp14 = tmp13 - tmp2;
                    auto tmp15 = tmp14 * tmp4;
                    tmp15.store(out_ptr0 + static_cast<long>(i1 + (784L*i0)));
                }
            }
        }
    }
}
''')
```

**Test Plan**
```
clear && python -m pytest test_cpu_repro.py -k test_non_contiguous_load_buf_quant
```

Pull Request resolved: https://github.com/pytorch/pytorch/pull/106958
Approved by: https://github.com/jgong5, https://github.com/eellison
ghstack dependencies: #106836, #106838

leslie-fang-intel committed:
[Quant][PT2E]Make _fuse_conv_bn_ support graph capture by torch._dynamo.export (#107951)

**Summary**
The latest check-in https://github.com/pytorch/pytorch/commit/a0cfaf068887a608c00de4f5ceaf30242c6af720 for the conv-bn folding assumes the graph is captured by the new graph capture API `torch._export.capture_pre_autograd_graph`. Since we still need to use the original graph capture API `torch._dynamo_export` in 2.1 release. So, this check-in made negative impact to workloads' performance heavily. Made this PR to fix this issue by trying to make the conv-bn folding function workable with both new and original graph capture API.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/107951
Approved by: https://github.com/jgong5, https://github.com/jerryzh168
ghstack dependencies: #106836, #106838, #106958

hl475 committed:
[dynamo+aten] Enable embedding_bag_byte_unpack + meta kernel impl (#107937)

Summary:
```
torch._dynamo.exc.Unsupported: unsupported operator: quantized.embedding_bag_byte_unpack.default
```

Differential Revision: D48652953

Pull Request resolved: https://github.com/pytorch/pytorch/pull/107937
Approved by: https://github.com/houseroad

valentinandrei committed:
[cuda] remove dead CUDA code in layer_norm_kernel.cu (#107976)

Removing CUDA kernels which are not used anywhere.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/107976
Approved by: https://github.com/Skylion007

mlazos committed:
Handle 2D blocking with foreach (#107840)

Previously blocking in foreach ops was only 1D. This PR allows handling kernels with 2D blocking with foreach as well.

Code when at least one dim matches:
[example code + output](https://gist.github.com/mlazos/9f100b21cfe2540f0a24303a8349c196)

Code when neither X or Y dim matches:
[example code + output](https://gist.github.com/mlazos/14e2a455f635896dface09be601595dd)

Pull Request resolved: https://github.com/pytorch/pytorch/pull/107840
Approved by: https://github.com/jansel

zhxchen17 committed:
[export] Don't save example_inputs for now. (#107978)

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Fixes #ISSUE_NUMBER

Pull Request resolved: https://github.com/pytorch/pytorch/pull/107978
Approved by: https://github.com/angelayi

jerryzh168 committed:
[quant][pt2e] Add reference representation rewrite for statically quantized linear (#107994)

Summary: att

Test Plan:
```
python test/test_quantization.py TestQuantizePT2E.test_representation_linear
buck2 test 'fbcodemode/opt' fbcodecaffe2/test:quantization_pt2e -- 'test_representation_linear'
```

Reviewed By: kimishpatel

Differential Revision: D48674862

Pull Request resolved: https://github.com/pytorch/pytorch/pull/107994
Approved by: https://github.com/mcr229, https://github.com/guangy10

abock committed:
[ONNX] Support `torch.compile(backend="onnxrt", options=OrtBackendOptions(...))` (#107973)

This reworks the DORT backend factory function to support the options kwarg of torch.compile, and defines a concrete OrtBackendOptions type that can be used to influence the backend.

Caching is also implemented in order to reuse backends with equal options.

Wrapping the backend in auto_autograd also becomes an option, which allows `OrtBackend` to always be returned as the callable for torch.compile; wrapping happens internally if opted into (True by default).

Lastly, expose options for configuring preferred execution providers (will be attempted first), whether or not to attempt to infer an ORT EP from a torch found device in the graph or inputs, and finally the default/fallback EPs.

### Demo

The following demo runs `Gelu` through `torch.compile(backend="onnxrt")` using various backend options through a dictionary form and a strongly typed form. It additionally exports the model through both the ONNX TorchScript exporter and the new TorchDynamo exporter.

```python
import math

import onnx.inliner
import onnxruntime
import torch
import torch.onnx

torch.manual_seed(0)

class Gelu(torch.nn.Module):
    def forward(self, x):
        return x * (0.5 * torch.erf(math.sqrt(0.5) * x) + 1.0)

@torch.compile(
    backend="onnxrt",
    options={
        "preferred_execution_providers": [
            "NotARealEP",
            "CPUExecutionProvider",
        ],
        "export_options": torch.onnx.ExportOptions(dynamic_shapes=True),
    },
)
def dort_gelu(x):
    return Gelu()(x)

ort_session_options = onnxruntime.SessionOptions()
ort_session_options.log_severity_level = 0

dort_gelu2 = torch.compile(
    Gelu(),
    backend="onnxrt",
    options=torch.onnx._OrtBackendOptions(
        preferred_execution_providers=[
            "NotARealEP",
            "CPUExecutionProvider",
        ],
        export_options=torch.onnx.ExportOptions(dynamic_shapes=True),
        ort_session_options=ort_session_options,
    ),
)

x = torch.randn(10)

torch.onnx.export(Gelu(), (x,), "gelu_ts.onnx")

export_output = torch.onnx.dynamo_export(Gelu(), x)
export_output.save("gelu_dynamo.onnx")
inlined_model = onnx.inliner.inline_local_functions(export_output.model_proto)
onnx.save_model(inlined_model, "gelu_dynamo_inlined.onnx")

print("Torch Eager:")
print(Gelu()(x))

print("DORT:")
print(dort_gelu(x))
print(dort_gelu2(x))
```

Pull Request resolved: https://github.com/pytorch/pytorch/pull/107973
Approved by: https://github.com/BowenBao

jataylo committed:
[ROCm] ROCm compatible configs for triton kernels (#107584)

This PR brings in a few inductor changes required for ROCm

~**1 - Introduction of a toggle for enforced channel last convolution fallbacks**~
This addition is split off into its own PR after some cleanup by @pragupta  https://github.com/pytorch/pytorch/pull/107812

**2 - Addition of ROCm specific block sizes**
We are now able to support the MAX_AUTOTUNE mode on ROCm, we are proposing conditions to allow us to finetune our own block tuning. Currently triton on ROCm does not benefit from pipelining so we are setting all configs to `num_stages=1` and we have removed some upstream tunings on ROCm to avoid running out of shared memory resources.

In the future we will provide more optimised tunings for ROCm but for now this should mitigate any issues

~**3 - Addition of device_type to triton's compile_meta**~
~Proposing this addition to `triton_heuristics.py`, Triton on ROCm requires device_type to be set to hip https://github.com/ROCmSoftwarePlatform/triton/pull/284 suggesting to bring this change in here so we can pass down the correct device type to triton.~
This change is split off and will arrive in the wheel update PR https://github.com/pytorch/pytorch/pull/107600 leaving this PR to focus on the ROCm specific block sizes.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/107584
Approved by: https://github.com/jithunnair-amd, https://github.com/jansel, https://github.com/eellison

james-a-watson committed:
Update writing_batching_rules.md (#108007)

Was reading through the batching rules info which is very cool and just saw a couple of typos 😊.

Thanks

Pull Request resolved: https://github.com/pytorch/pytorch/pull/108007
Approved by: https://github.com/msaroufim

pytorchupdatebot committed:
[vision hash update] update the pinned vision hash (#108011)

This PR is auto-generated nightly by [this action](https://github.com/pytorch/pytorch/blob/main/.github/workflows/_update-commit-hash.yml).
Update the pinned vision hash.
Pull Request resolved: https://github.com/pytorch/pytorch/pull/108011
Approved by: https://github.com/pytorchbot

voznesenskym committed:
WIP

WIP

WIP

WIP

WIP

WIP

WIP

voznesenskym commented:
mashed in the wrong direction, disregard https://github.com/pytorch/pytorch/pull/108018
